### ISSUE TITLE
Minimize single residue

### ISSUE THREAD

**Issue by jnoske** [openmm/openmm#3178]
Hello, how could I minimize only a single residue and keep the rest of the protein rigid, I also would like to keep the side chain torsions of this residue constrained, essentially the backbone should be minimized and the side chain should move. If I use external forces or set masses to zero the protein is always moving, maybe this is an over constrained system and not possible.

Thank you

**Comment by peastman** [comment]
Am I correctly understanding what you want?  The goal is that

- All atoms outside the one residue should remain immobile.
- The backbone atoms in the residue should move freely.
- The sidechain should move as a rigid body.

If that's correct, you'll need to add constraints to the sidechain to make it rigid.  That takes a bit of care to do correctly.  You need to add just the right number of constraints to remove all internal degrees of freedom, but not overconstrain anything.

**Comment by jnoske** [comment]
Exactly, the problem is, that if I add external forces or set the masses of all atoms not belonging to this residue to zero, some atoms still move during the minimization:

This:

for i, atom_crd in enumerate(inpcrd.positions):
    if structure.resid[i] != resid and structure.chain[i] != chain:
        system.setParticleMass(i, 0.0)

Or like this:

force = mm.CustomExternalForce("k*((x-x0)^2+(y-y0)^2+(z-z0)^2)")
force.addGlobalParameter("k", 5.0 * unit.kilocalories_per_mole / unit.angstroms ** 2)
force.addPerParticleParameter("x0")
force.addPerParticleParameter("y0")
force.addPerParticleParameter("z0")
for i, atom_crd in enumerate(inpcrd.positions):
    if structure.name[i] in ('CA', 'C', 'N'):
        force.addParticle(i, atom_crd.value_in_unit(unit.nanometers))
system.addForce(force)

How would I constrain the side chain torsions, to obtain a rigid body:

  force_sidechain = mm.CustomTorsionForce("0.5*k*min(dtheta, 2*pi-dtheta)^2; dtheta = abs(theta-theta0); pi = 3.1415926535")
  force_sidechain.addGlobalParameter("k", 5.0 * unit.kilocalories_per_mole / unit.radians ** 2)
  force_sidechain.addPerTorsionParameter("theta0")
  for torsion in _SIDECHAIN_TORSIONS:
      selection = f'resid {resid} and chain {chain} and (name '
      for i, atom in enumerate(torsion):
          if i < 3:
              selection += f'{atom} or name '
          else:
              selection += f'{atom})'
      indices = structure.get('index', selection)
      indices = [int(index) for index in indices]
      force_sidechain.addTorsion(*indices, [structure.getDihedral(indices)])
  system.addForce(force_sidechain)

Thanks, for the help !!

**Comment by peastman** [comment]
Are the atoms moving by only a tiny amount or a lot?  Setting masses to 0 should keep them fixed right in place.  A CustomExternalForce just limits how far they're able to move.  The larger the force constant, the less they can move.

Your torsion restraint is different from a rigid body.  You're leaving all the bonds and angles unconstrained, and just limiting how far the dihedrals can move.  The side chain will still be flexible, but with its flexibility limited by the restraining force.  If that's good enough for your purposes, that's *much* easier than actually making it rigid.  Does the code above work?

**Comment by jnoske** [comment]
Ok, so I would like to restrain only the torsion angles to keep the current rotamer. Furthermore, I want to make the rest of the protein completely rigid, however even after setting the masses of all atoms to zero, there are movements in other side chains. I would like to minimize the energy of a residue with constrained torsions in a rigid protein.

**Comment by peastman** [comment]
Can you verify that the side chains that are moving really have their masses set to 0?  The code above gets atom indices from an object called `structure`.  It doesn't look like any kind of OpenMM object.  Could there be a mismatch such that the wrong atoms are getting fixed?

When LocalEnergyMinimizer communicates the gradients to the LBFGS library, it sets them to 0 for any massless particle, so it really shouldn't be moving them:

https://github.com/openmm/openmm/blob/c74510f7f492087f24d43754e8a7d2ee1d0790ee/openmmapi/src/LocalEnergyMinimizer.cpp#L85-L96

**Comment by jnoske** [comment]
Yes you are right, I had a problem with selecting all the indices. I can set now the entire protein rigid, the problem is, still that even after constraining the torsion angles like described above. That now after minimization all initial different rotamers have the same coordinates, however I want to keep torsions contrained and mainly want to align the backbone.

**Comment by jnoske** [comment]
Thank you, now the protein is completely rigid, however I still have problems with constraining torsion angles. In the end all rotamers of my residue have the same torsion angles and coordinates. Is my way of constraining torsion angles correct and is it again just a problem with wrong indices ?

**Comment by peastman** [comment]
> In the end all rotamers of my residue have the same torsion angles and coordinates.

I don't understand what you mean.  The code above sets a single restraint on each torsion.  Do you mean you're running the code multiple times, setting the restraints to different angles each time, to sample different rotamers?

**Comment by jnoske** [comment]
No I have different rotamers of my residue and I want to minimize them, but with keeping their torsion angles which are different.

**Comment by peastman** [comment]
Then you'll set the per-torsion parameters to different values for each rotamer.  I don't understand what question you're asking.

**Comment by jnoske** [comment]
I want to ask if this, is the right way to do it:

force_sidechain = mm.CustomTorsionForce("0.5kmin(dtheta, 2*pi-dtheta)^2; dtheta = abs(theta-theta0); pi = 3.1415926535")
force_sidechain.addGlobalParameter("k", 5.0 * unit.kilocalories_per_mole / unit.radians ** 2)
force_sidechain.addPerTorsionParameter("theta0")
for torsion in _SIDECHAIN_TORSIONS:
selection = f'resid {resid} and chain {chain} and (name '
for i, atom in enumerate(torsion):
if i < 3:
selection += f'{atom} or name '
else:
selection += f'{atom})'
indices = structure.get('index', selection)
indices = [int(index) for index in indices]
force_sidechain.addTorsion(*indices, [structure.getDihedral(indices)])
system.addForce(force_sidechain)

And is their a way of getting the indices of the particles in the system, since I obtain indices from a moleculekit.molecule.Molecule object to restrain the torsion angles.

**Comment by peastman** [comment]
You have a per-torsion parameter `theta0` that sets the target angle for each restraint.  You set the value of it to `structure.getDihedral(indices)`.  If you want to do another minimization with a different target angle, you need to change the value of that parameter.

> And is their a way of getting the indices of the particles in the system, since I obtain indices from a moleculekit.molecule.Molecule object to restrain the torsion angles.

I'm not familiar with moleculekit.  It looks like it numbers things differently than OpenMM does.  How do you create the System?  Do you build it entirely yourself, or is it built by an OpenMM class like ForceField or AmberPrmtopFile?  If the latter, that means you also have a Topology for it.  The Topology numbers atoms in the same way the System does.
